src/cmd/board.rs

```diff
@@ -0,0 +1,96 @@
+// Copyright 2023 The ChromiumOS Authors
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file or at
+// https://developers.google.com/open-source/licenses/bsd
+
+use anyhow::bail;
+use anyhow::Result;
+use argh::FromArgs;
+use glob::Pattern;
+use lium::cache::KvCache;
+use lium::repo::get_repo_dir;
+use lium::util::shell_helpers::run_bash_command;
+
+#[derive(FromArgs, PartialEq, Debug)]
+/// list supported boards
+#[argh(subcommand, name = "board")]
+pub struct Args {
+    #[argh(subcommand)]
+    nested: SubCommand,
+}
+
+pub static BOARD_CACHE: KvCache<Vec<String>> = KvCache::new("board_cache");
+
+#[derive(FromArgs, PartialEq, Debug)]
+#[argh(subcommand)]
+enum SubCommand {
+    List(ArgsList),
+}
+pub fn run(args: &Args) -> Result<()> {
+    match &args.nested {
+        SubCommand::List(args) => run_board_list(args),
+    }
+}
+
+#[derive(FromArgs, PartialEq, Debug)]
+/// List up the supported boards
+#[argh(subcommand, name = "list")]
+pub struct ArgsList {
+    /// target cros repo directory
+    #[argh(option)]
+    repo: Option<String>,
+
+    /// glob pattern of the boards
+    #[argh(positional)]
+    filter: Option<String>,
+
+    /// only show cached list
+    #[argh(switch)]
+    cached: bool,
+}
+
+fn print_cached_boards(filter: &Pattern) -> Result<()> {
+    let boards = BOARD_CACHE.entries()?;
+    if boards.is_empty() {
+        bail!("No cache found");
+    }
+
+    let mut boards: Vec<String> = boards.into_keys().collect();
+    boards.sort();
+
+    for board in boards {
+        if filter.matches(&board) {
+            println!("{board}");
+        }
+    }
+    Ok(())
+}
+
+fn update_cached_boards(repodir: &str) -> Result<()> {
+    let output = run_bash_command("cros query boards", Some(repodir))?;
+    output.status.exit_ok()?;
+    let list = String::from_utf8(output.stdout)?;
+    for s in list.lines() {
+        BOARD_CACHE.set(s, Vec::<String>::new())?;
+    }
+    Ok(())
+}
+
+fn run_board_list(args: &ArgsList) -> Result<()> {
+    let filter = if let Some(_filter) = &args.filter {
+        Pattern::new(_filter)?
+    } else {
+        Pattern::new("*")?
+    };
+
+    if print_cached_boards(&filter).is_ok() || args.cached {
```

Comment:
it looks like the current logic does not update the cached boards when it is cached already. how about just skipping the update phase when the --cached is specified?

Reply:
Good catch! OK.

Reply:
Note that this changes the default behavior a bit.
`lium board list` will always query the boards unless `--cached` is specified. (previously, even without `--cached`, this shows the cached list and does not query it again.)